napari/layers/image/image.py

```diff
@@ -689,3 +689,34 @@ def contrast_limits(self, contrast_limits):
                 self.refresh()
             finally:
                 self._keep_auto_contrast = prev
+
+    def _calculate_value_from_ray(self, values):
+        values_within_clim = (values >= self.contrast_limits[0]) & (
+            values <= self.contrast_limits[1]
+        )
+        values_masked = np.where(values_within_clim, values, np.nan)
+        if np.all(np.isnan(values_masked)):
+            return None  # out of clims everywhere
+        if self.rendering == ImageRendering.TRANSLUCENT:
+            first_val = np.ravel(values_masked)[0]
+            return first_val if not np.isnan(first_val) else None
+        # these are probably not the same as how the gpu does it...
+        if self.rendering == ImageRendering.AVERAGE:
+            return np.nanmean(values_masked)
+        if self.rendering == ImageRendering.ADDITIVE:
+            # TODO: this is "broken" cause same pixel gets multisampled...
+            #       but it looks like it's also overdoing it in vispy vis too?
+            #       I don't know if there's a way to *not* do it...
+            return np.nansum(values_masked)
```

Comment:
Not sure if there is any way around it. See the test, we get a `2` as value because of oversampling, where the "sensible" result would be 1... but then again, switching from `mip` to `additive` rendering *does* result in an increase in brightness, so that's consistent.

Reply:
What about clamping the output to the contrast limits here? actually that may need to apply to all. I think the shader generally has early termination for fully-saturated values so this makes sense to me.

Reply:
I don't think it's the case for additive; Though I must say I don't understand the shader code:

```glsl
color = applyTransferFunction(color, loc, step, depth_origin, depth_plane_normal, max_depth);

integrated_color = 1.0 - (1.0 - integrated_color) * (1.0 - color);
```

Reply:
For the others I think that's correct, but here you're summing many values that could all be the max contrast limit for example.

The shader is not doing a simple sum, but blending colors. Thus I think it will converge a bit weirdly, depending on the maximum of each channel in the colormap. If the colormap is a simple colormap like "grays" it will converge to `applyColormap(clim.y)`. I think this is the intent, so I would recommend clamping after the sum here.

There might be a better formula, and this probably could/should do early termination in the shader when `1.0 - integrated_color` is close enough to 0 (or something), but I don't think it will make a difference in the resulting color.

> As for the rest, clims should always be respected because I'm masking the out-of-range values in the first lines of the method.

This is mostly correct but there's a problem there as well. For example you can get a value of `0` on a MIP by bringing the high contrast limit down. Maybe you only want to mask out-of-range values below the low contrast limit?

Reply:
I'm starting to think it makes no sense to try and get the correct displayed value... Doesn't the user actually want to know what the value of the *data* under the cursor is?

> This is mostly correct but there's a problem there as well. For example you can get a value of 0 on a MIP by bringing the high contrast limit down. Maybe you only want to mask out-of-range values below the low contrast limit?

This is intended: if your contrast limit is low enough that nothing is displayed in that pixel, then you should see nothing? Also, you don't get `0`, but you get `None`.

Reply:
> Doesn't the user actually want to know what the value of the data under the cursor is?

👍 - whatever this may mean in a volume rendering context


> This is intended: if your contrast limit is low enough that nothing is displayed in that pixel, then you should see nothing? Also, you don't get 0, but you get None.

But you're not seeing nothing - you see bright white (in the case of MIP with gray colormap).

Reply:
To me it would mean *the value of the data at the position where the color displayed belongs to*. So, for additive, nothing xD @jni opinions?

I see, you're right... Something's wrong it seems. I think I need to work on this more, I'll move to 0.5.3.

Reply:
> To me it would mean the value of the data at the position where the color displayed belongs to

I agree with this. But I don't understand this:

> So, for additive, nothing xD

til about additive *rendering* instead of additive *blending* 😂 I agree that it's tricky. I think what users want is position (in data space) + value. What we could do for the "summary" renderings (additive, average) is to show the ray start *and finish*, followed by the value.

I don't think it's a huge deal if the value is *slightly* off due to oversampling because floating points are weird.

Reply:
(But I'm happy to have a bit more time to get this right until 0.5.3)